sympy/matrices/tests/test_matrices.py

```diff
@@ -222,31 +222,32 @@ def test_power():
         [0, 0, b**n]])
 
     A = Matrix([[1, 0], [1, 7]])
-    assert A._matrix_pow_by_jordan_blocks(3) == A._eval_pow_by_recursion(3)
+    assert A._matrix_pow_by_jordan_blocks(S(3)) == A._eval_pow_by_recursion(3)
     A = Matrix([[2]])
-    assert A**10 == Matrix([[2**10]]) == A._matrix_pow_by_jordan_blocks(10) == \
+    assert A**10 == Matrix([[2**10]]) == A._matrix_pow_by_jordan_blocks(S(10)) == \
         A._eval_pow_by_recursion(10)
 
     # testing a matrix that cannot be jordan blocked issue 11766
     m = Matrix([[3, 0, 0, 0, -3], [0, -3, -3, 0, 3], [0, 3, 0, 3, 0], [0, 0, 3, 0, 3], [3, 0, 0, 3, 0]])
-    raises(MatrixError, lambda: m._matrix_pow_by_jordan_blocks(10))
+    raises(MatrixError, lambda: m._matrix_pow_by_jordan_blocks(S(10)))
 
     # test issue 11964
-    raises(ValueError, lambda: Matrix([[1, 1], [3, 3]])._matrix_pow_by_jordan_blocks(-10))
+    raises(ValueError, lambda: Matrix([[1, 1], [3, 3]])._matrix_pow_by_jordan_blocks(S(-10)))
     A = Matrix([[0, 1, 0], [0, 0, 1], [0, 0, 0]])  # Nilpotent jordan block size 3
     assert A**10.0 == Matrix([[0, 0, 0], [0, 0, 0], [0, 0, 0]])
     raises(ValueError, lambda: A**2.1)
     raises(ValueError, lambda: A**(S(3)/2))
     A = Matrix([[8, 1], [3, 2]])
     assert A**10.0 == Matrix([[1760744107, 272388050], [817164150, 126415807]])
     A = Matrix([[0, 0, 1], [0, 0, 1], [0, 0, 1]])  # Nilpotent jordan block size 1
-    assert A**10.2 == Matrix([[0, 0, 1], [0, 0, 1], [0, 0, 1]])
+    assert A**10.0 == Matrix([[0, 0, 1], [0, 0, 1], [0, 0, 1]])
     A = Matrix([[0, 1, 0], [0, 0, 1], [0, 0, 1]])  # Nilpotent jordan block size 2
     assert A**10.0 == Matrix([[0, 0, 1], [0, 0, 1], [0, 0, 1]])
     n = Symbol('n', integer=True)
     assert isinstance(A**n, MatPow)
-    n = Symbol('n', integer=True, nonnegative=True)
+    n = Symbol('n', integer=True, negative=True)
     raises(ValueError, lambda: A**n)
+    n = Symbol('n', integer=True, nonnegative=True)
     assert A**(n + 2) == Matrix([[0, 0, 1], [0, 0, 1], [0, 0, 1]])
```

Comment:
Can we also check `A**n` here?

Reply:
Its an ugly and long specific value matrix that would be wrong if someone changes the implementation to something else, the important test here is the logical one that this matrix to any power >= 2 will be this.

Reply:
Or any power >= 1? I don't think it's that ugly:
```julia
In [5]: M**n                                                                                             
Out[5]: 
⎡δ      0    1 - δ   ⎤
⎢ 0,n             0,n⎥
⎢                    ⎥
⎢ 0    δ     1 - δ   ⎥
⎢       0,n       0,n⎥
⎢                    ⎥
⎣ 0     0       1    ⎦
```
In [5]: M**n                                                                                             
Out[5]: 
⎡δ      0    1 - δ   ⎤
⎢ 0,n             0,n⎥
⎢                    ⎥
⎢ 0    δ     1 - δ   ⎥
⎢       0,n       0,n⎥
⎢                    ⎥
⎣ 0     0       1    ⎦
```
You've added this KroneckerDelta code to get the correct answer in the most general situation possible. That needs to be tested.